src/sidebar/components/thread-list-omega.js

```diff
@@ -0,0 +1,246 @@
+  const MARGIN_ABOVE = 800;
+  // Same as MARGIN_ABOVE but for the space below the viewport
+  const MARGIN_BELOW = 800;
+  // List of threads which are in or near the viewport and need to
+  // actually be rendered.
+  const visibleThreads = [];
+
+  const calculatedHeights = {
+    offscreenLower: 0,
+    offscreenUpper: 0,
+    total: 0,
+  };
+
+  const threadIsAboveViewport = (threadHeight, heights) =>
+    heights.total + threadHeight < scrollPos - MARGIN_ABOVE;
+  const threadIsVisible = (threadHeight, heights) =>
+    heights.total < scrollPos + windowHeight + MARGIN_BELOW;
+
+  threads.forEach(thread => {
+    const threadHeight = threadHeights[thread.id] || THREAD_DEFAULT_HEIGHT;
+    if (threadIsAboveViewport(threadHeight, calculatedHeights)) {
+      calculatedHeights.offscreenUpper += threadHeight;
+    } else if (threadIsVisible(threadHeight, calculatedHeights)) {
+      visibleThreads.push(thread);
+    } else {
+      // thread is below visible viewport
+      calculatedHeights.offscreenLower += threadHeight;
+    }
+    calculatedHeights.total += threadHeight;
+  });
+
+  return {
+    visibleThreads,
+    offscreenUpperHeight: calculatedHeights.offscreenUpper,
+    offscreenLowerHeight: calculatedHeights.offscreenLower,
+  };
+}
+
+function getScrollContainer() {
+  return document.querySelector('.js-thread-list-scroll-root');
+}
+
+/**
+ * Render a list of threads, but only render those that are in or near the
+ * current browser viewport.
+ */
+function ThreadListOmega({ thread, $rootScope }) {
+  const clearSelection = useStore(store => store.clearSelection);
+
+  // Height of the scrollable container. For the moment, this is the same as
+  // the window height.
+  const [windowHeight, setWindowHeight] = useState(window.innerHeight);
+
+  // Scroll offset of scroll container. This is updated after the scroll
+  // container is scrolled, with debouncing to limit update frequency.
+  const [scrollPosition, setScrollPosition] = useState(
+    getScrollContainer().scrollTop
+  );
+
+  // Map of thread ID to measured height of thread.
+  const [threadHeights, setThreadHeights] = useState({});
+
+  // ID of thread to scroll to after the next render. If the thread is not
+  // present, the value persists until it can be "consumed".
+  const [scrollToId, setScrollToId] = useState(null);
+
+  // Effect to scroll to a new annotation when it is created.
+  useEffect(() => {
+    // When a new annotation is created in the application, scroll to that
+    // annotation's thread (unless it is a highlight or reply)
+    $rootScope.$on(events.BEFORE_ANNOTATION_CREATED, (event, annotation) => {
+      if (!isHighlight(annotation) && !isReply(annotation)) {
+        clearSelection();
+
+        // If the annnotation is already visible, scroll directly to it, otherwise
+        // we'll need to calculate its _expected_ position and scroll there.
+        setScrollToId(annotation.$tag);
+      }
+    });
```

Comment:
We're missing logic to clean up this effect properly here, although in practice `$rootScope` and `clearSelection` will never change.